src/raftstore/store/worker/consistency_check.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::collections::HashMap;
+use std::fmt::{self, Formatter, Display};
+
+use crc::crc32::{self, Digest, Hasher32};
+use byteorder::{BigEndian, WriteBytesExt};
+
+use kvproto::metapb::Region;
+use raftstore::store::{keys, Msg};
+use raftstore::store::engine::{Snapshot, Iterable, Peekable};
+use storage::CF_RAFT;
+use util::escape;
+use util::worker::Runnable;
+
+use super::metrics::*;
+use super::MsgSender;
+
+/// Consistency checking task.
+pub enum Task {
+    ComputeHash {
+        index: u64,
+        region: Region,
+        snap: Snapshot,
+    },
+    VerifyHash {
+        region_id: u64,
+        index: u64,
+        hash: Vec<u8>,
+    },
+}
+
+impl Task {
+    pub fn compute_hash(region: Region, index: u64, snap: Snapshot) -> Task {
+        Task::ComputeHash {
+            region: region,
+            index: index,
+            snap: snap,
+        }
+    }
+
+    pub fn verify_hash(region_id: u64, index: u64, hash: Vec<u8>) -> Task {
+        Task::VerifyHash {
+            region_id: region_id,
+            index: index,
+            hash: hash,
+        }
+    }
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            Task::ComputeHash { ref region, index, .. } => {
+                write!(f, "Compute Hash Task for {:?} at {}", region, index)
+            }
+            Task::VerifyHash { region_id, index, ref hash } => {
+                write!(f,
+                       "Verify Hash Task for {} at {} [hash: {}]",
+                       region_id,
+                       index,
+                       escape(hash))
+            }
+        }
+    }
+}
+
+pub struct Runner<C: MsgSender> {
+    ch: C,
+    // region_id -> (index, checksum)
+    checksum: HashMap<u64, (u64, u32)>,
```

Comment:
Runner is not designed to be stateless, there are several use cases that keep state inside runner, like endpoint, resolve worker and Region worker. But it's better to check checksum in store thread.